### ISSUE TITLE
[Layout] Disallow dropping the same object

### ISSUE THREAD

**Issue by psarram** [nasa/openmct#1815]
Currently one object per layout is supported, but dropping the same object is allowed.

To reproduce:
- Drop an object to a layout. It gets positioned and selected. 
- Deselect the object.
- Drop the same object again. It gets placed in a new position but remains unselected.

**Comment by akhenry** [comment]
TL;DR - We should always do something rather than nothing on drop.

My understanding of this issue is that
* Layouts may contain multiple objects, but may only contain any given object once. 
* An object that is already contained in a layout can currently be dropped into that layout, and the effect is to 'move' the object to the drop position, and deselect it.

Before doing anything here, I'd like to pause and think about the desired behavior. The effect of moving the object to a new location when it's dropped into a layout that already contains it is disconcerting at times. Most of the time I end up moving it back to where it was beforehand, so it's probably not the desired behavior. That said, it actually provides some feedback to the user. It has the (perhaps unintended) effect of bringing to your attention to the object, and indicates that the drop did something. 

My concern is that if nothing happens when the object is dropped, then the user will be confused. Why didn't the drop work? Is it because the layout can't contain this object, or because the object is already in the layout, or some other reason?

In the case that a layout already contains an object, could we highlight it somehow? Maybe flash the border for a second or so? A simple way to do this would be to always flash the border on drop into a layout. It may also require scrolling to the object if it's off-screen though...

An alternative would be to show an info message in the bottom bar, but my preference is to visually direct the user to where the existing object is in the layout.

A corollary to this is that when a layout cannot contain a dropped object (eg. a folder) then we should indicate this somehow. Either visually (eg. a different cursor on drag over, or some style change in the layout or dragged object), or with an error message on drop.

**Comment by larkin** [comment]
I would argue that doing nothing on drop may be correct, but the UI should better indicate that nothing will happen on drop. 

There isn’t much we can do to improve this at this time, we have other changes that will invalidate this.

I’m down to close this.

**Comment by lopezdp** [comment]
Hey guys, I interned over at JSC last summer building out the web app related to the NASA/CCDD project and was able to get it working so it could be published as part of Core Flight Software. Unfortunately although it’s been approved for Open Source it will not be out on Git for another year while they wait for USRA’s decision to patent it or not (which they won’t), and unfortunately it means I’m short a project to contribute to. Getting these threads everyday in my inbox is like a siren calling me into the storm and I’m loving it although I’ve read so little of the docs.

Anyways, guys what I’m getting at is that this is a really cool project to contribute to.

You guys got any insight for a former NASA intern whose wandering the current planet we’re on looking for the next project to work on?

What’s a good issue to work on right now that’s holding you up? Maybe you got something bugging you (pun intended) on this one that I could write a quick test for or something just to get my feet wet since I’m feeling a little timid watching all you jedi’s Kill it on mission control software?

Seriously though, who the hell even writes mission control software anyway?

Im a fan of tech humor by the way.

How are you guys anyway, my name is LopezDP! :)

**Comment by akhenry** [comment]
> I would argue that doing nothing on drop may be correct, but the UI should better indicate that nothing will happen on drop.

I would argue that we're arguing the same thing :)

**Comment by larkin** [comment]
I would argue that my argument was in support of your argument! 

...except that it should be clear that nothing will happen **_before_** they drop.  The UI should provide hints to the user of valid drop targets (and the expected action) while the user is dragging-- not as a result of the drop.

**Comment by akhenry** [comment]
I would argue that we're in agreement then, and that no further argument is necessary :)

My only addendum is that it would be nice to be able to indicate to the user why the drop won't work, to differentiate between containment type policy violation, and potential duplication.

**Comment by akhenry** [comment]
@lopezdp Hey! Thanks for your interest in Open MCT. We have a bunch of Help Wanted issues in our backlog - https://github.com/nasa/openmct/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22

Specifically, https://github.com/nasa/openmct/issues/1872 is a really minor but annoying bug, if you're interested. Let me know if you're interested in that one, because otherwise one of us will probably fix it when we get a moment in the next couple of days, but we'll hold off if you're working on it.

For a more involved issue, there is this feature request -
 https://github.com/nasa/openmct/issues/354 . We just never seem to have the time to do this one, but it's __super__ annoying. If you did that one, we would build a statue in your honor.

Also, take a look at the [contributor guide](https://nasa.github.io/openmct/contributing/). We need a signed Contributor License Agreement before we can merge anything you do.

**Comment by charlesh88** [comment]
Fixed.
